### PR TITLE
Primitives

### PR DESCRIPTION
**Pull request by Licini**
Adding the all the rest of primitives
image

### CODE REVIEW

File: src/compas_view2/objects/capsuleobject.py [hunk 1]
```diff
@@ -0,0 +1,12 @@
+from compas.datastructures import Mesh
+from .meshobject import MeshObject
+
+
+class CapsuleObject(MeshObject):
+    """Object for displaying COMPAS cone geometry."""
```
Comment (brgcode) [inline src/compas_view2/objects/capsuleobject.py:6]: COMPAS capsule geometry

File: src/compas_view2/objects/circleobject.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+from .polylineobject import PolylineObject
+from compas.geometry import Frame
+import math
+
+
+class CircleObject(PolylineObject):
+    """Object for displaying COMPAS Plane geometry."""
```
Comment (brgcode) [inline src/compas_view2/objects/circleobject.py:7]: circle geometry

File: src/compas_view2/objects/compositeobject.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+from .bufferobject import BufferObject
+import numpy as np
+
+
+class CompositeObject(BufferObject):
+    """Object for displaying a composition of compas_view2 objects."""
```
Comment (brgcode) [inline src/compas_view2/objects/compositeobject.py:6]: personal preference: composition of View2 objects

File: src/compas_view2/objects/ellipseobject.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+from .polylineobject import PolylineObject
+from compas.geometry import Frame
+import math
+
+
+class EllipseObject(PolylineObject):
+    """Object for displaying COMPAS Plane geometry."""
```
Comment (brgcode) [inline src/compas_view2/objects/ellipseobject.py:7]: ellipse geometry

File: src/compas_view2/objects/meshobject.py [hunk 5]
```diff
@@ -238,7 +189,10 @@ def _frontfaces_data(self):
     def _backfaces_data(self):
         mesh = self._mesh
         vertex_xyz = {vertex: mesh.vertex_attributes(vertex, 'xyz') for vertex in mesh.vertices()}
-        face_color = self.facecolor
+        face_color = {
+            face: self._mesh.face_attribute(face, 'color') or self.facecolor or self.default_color_faces
+            for face in self._mesh.faces()
+            }
```
Comment (brgcode) [inline src/compas_view2/objects/meshobject.py:195]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/meshobject.py [hunk 6]
```diff
@@ -184,7 +132,10 @@ def _lines_data(self):
     def _frontfaces_data(self):
         mesh = self._mesh
         vertex_xyz = {vertex: mesh.vertex_attributes(vertex, 'xyz') for vertex in mesh.vertices()}
-        face_color = self.facecolor
+        face_color = {
+            face: self._mesh.face_attribute(face, 'color') or self.facecolor or self.default_color_faces
+            for face in self._mesh.faces()
+            }
```
Comment (brgcode) [inline src/compas_view2/objects/meshobject.py:138]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/meshobject.py [hunk 7]
```diff
@@ -156,7 +101,10 @@ def _points_data(self):
     def _lines_data(self):
         mesh = self._mesh
         vertex_xyz = {vertex: mesh.vertex_attributes(vertex, 'xyz') for vertex in mesh.vertices()}
-        linecolor = self.linecolor
+        linecolor = {
+            edge: self._mesh.edge_attribute(edge, 'color') or self.linecolor or self.default_color_lines
+            for edge in self._mesh.edges()
+            }
```
Comment (brgcode) [inline src/compas_view2/objects/meshobject.py:107]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/meshobject.py [hunk 8]
```diff
@@ -86,61 +79,13 @@ def __init__(self, data, color=None,
...
-                    facecolor[face] = color
-
     def _points_data(self):
         mesh = self._mesh
         vertex_xyz = {vertex: mesh.vertex_attributes(vertex, 'xyz') for vertex in mesh.vertices()}
-        vertex_color = self.pointcolor
+        vertex_color = {
+            vertex: self._mesh.vertex_attribute(vertex, 'color') or self.pointcolor or self.default_color_points
+            for vertex in self._mesh.vertices()
+            }
```
Comment (brgcode) [inline src/compas_view2/objects/meshobject.py:88]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/planeobject.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            self.frame.to_world_coordinates([0, 0, size])
+            )
+        lineObject = LineObject(line, **kwargs)
+
+        vertices = [
+            self.frame.to_world_coordinates([-size, -size, 0]),
+            self.frame.to_world_coordinates([size, -size, 0]),
+            self.frame.to_world_coordinates([size, size, 0]),
+            self.frame.to_world_coordinates([-size, size, 0])
+            ]
```
Comment (brgcode) [inline src/compas_view2/objects/planeobject.py:28]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/planeobject.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    """Object for displaying COMPAS Plane geometry."""
+
+    def __init__(self, data, size=1, **kwargs):
+        self._data = data
+        self.frame = Frame.from_plane(data)
+
+        line = Line(
+            self.frame.to_world_coordinates([0, 0, 0]),
+            self.frame.to_world_coordinates([0, 0, size])
+            )
```
Comment (brgcode) [inline src/compas_view2/objects/planeobject.py:20]: personal preference: indent one less or include at end of previous line

File: src/compas_view2/objects/polyhedronobject.py [hunk 11]
```diff
@@ -0,0 +1,10 @@
+from compas.datastructures import Mesh
+from .meshobject import MeshObject
+
+
+class PolyhedronObject(MeshObject):
+    """Object for displaying COMPAS cone geometry."""
```
Comment (brgcode) [inline src/compas_view2/objects/polyhedronobject.py:6]: polyhedron geometry

File: src/compas_view2/objects/polylineobject.py [hunk 12]
```diff
@@ -7,10 +7,11 @@ class PolylineObject(BufferObject):
     default_color_points = [0.1, 0.1, 0.1]
     default_color_line = [0.4, 0.4, 0.4]
 
-    def __init__(self, data, pointcolor=None, linecolor=None, color=None, **kwargs):
+    def __init__(self, data, closed=False, pointcolor=None, linecolor=None, color=None, **kwargs):
```
Comment (brgcode) [inline src/compas_view2/objects/polylineobject.py:10]: what is the meaning of "closed"? from the snippets below it is not clear to me.

File: src/compas_view2/objects/polylineobject.py [hunk 13]
```diff
@@ -26,9 +27,13 @@ def _lines_data(self):
         positions = []
         colors = []
         elements = []
-        for i in range(len(polyline.points)-1):
+        if self.closed:
+            indexes = range(len(polyline.points))
+        else:
+            indexes = range(len(polyline.points)-1)
+        for i in indexes:
```
Comment (brgcode) [inline src/compas_view2/objects/polylineobject.py:34]: why not go through the indices in pairs using `pairwise`? seems to remove the need for the modulo operation...

Comment (Licini) [reply src/compas_view2/objects/polylineobject.py:34]: makes sense!

File: src/compas_view2/objects/polylineobject.py (hunk 12 above)
Comment (Licini) [reply src/compas_view2/objects/polylineobject.py:10]: this is to whether close the polyline by connecting start to end

Comment (brgcode) [reply src/compas_view2/objects/polylineobject.py:10]: should it then not be "close" rather than "closed"?

**Comment by Licini** [comment]
@brgcode All Adjusted

**Review by brgcode — APPROVED** [review]

**Pull request merged**